### PR TITLE
Better grouped convolution for CPU targets

### PR DESCRIPTION
**Pull request by Wheest**
This pull request is to replace the current grouped direct convolution algorithm on x86 and Arm targets, with the faster Grouped Spatial Pack Convolutions (GSPC) algorithm.

Here's a performance comparison graph for ResNet34 on a single big core of a Hikey 970 as we increase the number of groups:
tvm_PR_hikey_1thr_ResNet34_tex

Note that in the untuned case the current depthwise convolution outperforms GSPC, thus I have omitted it from the pull request.

This is my first proper full request to TVM, so I may be have some issues I haven't spotted, or style problems.

In short, this commit adds identical GSPC compute definitions and schedules for x86 and arm_cpu targets for grouped convolutions, as well as updating the Relay operator strategy for each.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @anijain2305 @FrozenGene @mbaret @giuseros it would be great if you can help to review the PR

**Comment by tqchen** [comment]
@Wheest please help to fix the CI lint error

File: topi/python/topi/arm_cpu/group_conv2d.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+    data_vec = te.compute(shape,
+                          lambda g, n, C, h, c, w:
+                          data_pad[n, C * ic_bn + c + CPG * g, h, w],
+                          name='data_vec')
+
+    # pack kernel
+    shape = (G, KPG//oc_bn, CPG//ic_bn,
+             KH, KW, ic_bn, oc_bn)
+    kernel_vec = te.compute(shape,
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/group_conv2d.py:148]: I think we could do this in alter_op_layout for kernel, then we won't need do schedule kernel_vec as it will become tensor when we do inference. Could refer spatial pack of arm conv2d.

File: topi/python/topi/arm_cpu/group_conv2d.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    HPAD, WPAD = padding
+    DOPAD = (HPAD != 0 or WPAD != 0)
+
+    A, W = data, kernel_vec
+    A0, A1 = data_pad, data_vec
+
+    # schedule data
+    if DOPAD:
+        s[A0].compute_inline()
+    groups, batch, ic_chunk, ih, ic_block, iw = s[A1].op.axis
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/group_conv2d.py:251]: Maybe we should have one compute_at for A1 at CC? Not only just parallel.

File: topi/python/topi/arm_cpu/group_conv2d.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    padding = infer_pad(data, data_pad)
+    HPAD, WPAD = padding
+    DOPAD = (HPAD != 0 or WPAD != 0)
+
+    A, W = data, kernel_vec
+    A0, A1 = data_pad, data_vec
+
+    # schedule data
+    if DOPAD:
+        s[A0].compute_inline()
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/group_conv2d.py:250]: Consider using compute_at and vectorize data load. According to experiment of https://github.com/apache/incubator-tvm/pull/6095#discussion_r458893850, we should have better performance than compute inline directly.

File: topi/python/topi/arm_cpu/group_conv2d.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    else:
+        dilation_h, dilation_w = dilation
+
+    assert isinstance(padding, int) or len(padding) == 2 or len(padding) == 4
+    if isinstance(padding, int):
+        HPAD, WPAD = padding, padding
+    elif len(padding) == 2:
+        HPAD, WPAD = padding
+    else:
+        HPAD, _, WPAD, _ = padding
```
[the remark this answers is not in the document]
Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py:93]: Have added this to my working version.  However, the support function [_get_workload()](https://github.com/apache/incubator-tvm/blob/master/python/tvm/topi/nn/conv2d.py#L141) throws this information away.

[the remark this answers is not in the document]
Comment (FrozenGene) [reply topi/python/topi/arm_cpu/group_conv2d.py:93]: I think we should extend `_get_workload` function to support this, do you think so?

[the remark this answers is not in the document]
Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py]: I've got a suggestion for extending _get_workload.  In the new commit `505c127` I've added a 2nd Workload, `Workload_asym`.  `_get_workload` takes an optional argument `asymmetric_pad` which means it will return this workload instead.  Ideally the old conv2d `Workload` would be deprecated so that all conv2d workloads support asymmetric padding.

**Comment by ZihengJiang** [comment]
Ping for update @FrozenGene @Wheest

**Comment by Wheest** [comment]
@ZihengJiang Have updated linting errors.  

Have added asymmetric padding support, however this isn't consistent across the rest of TOPI.   Have added a new workload type in `python/tvm/topi/nn/conv2d.py`, that can be adopted across other Conv2D implementations.

Have been looking at suggested scheduled improvements, however have not been able to get any improvements to date.

Have not yet figured out how to do the kernel packing step in `alter_op_layout`, but I think I can do when I have some time next week.

**Comment by ZihengJiang** [comment]
Sounds good! Just ping the reviewers when you feel it is ready to be reviewed again. Thanks for the works!

**Comment by Wheest** [comment]
Hello there, updating this pull request to be up-to-date with the latest `main` branch.

In terms of things remaining to do:

- [x] [Consider using compute_at and vectorize data load](https://github.com/apache/tvm/pull/6137#discussion_r463852066) - did not get an improvement.
- [x] [We should support asymmetic padding like other compute / schedule.](https://github.com/apache/tvm/pull/6137#pullrequestreview-459474121) - this is implemented in GSPC, however requires extending `get_workload` for Conv2D generally.  I began working on this in `505c127`, but have reverted it, and will have this as it's own pull request in the comings days.
- [ ] [Pack in alter_op_layout for kernel](https://github.com/apache/tvm/pull/6137#discussion_r463844394):  have been working on this, but have an issue.   My data is being passed to my `group_conv2d_NCHWc.x86` in the `conv2d_NCHWc` format (5D input data), rather than the GSPC format (6D input data).  Despite my changes to the x86 `_alter_conv2d_layout`.  [See this branch](https://github.com/Wheest/incubator-tvm/tree/wheest/gspc-dev-alter-op).  Some guidance or pointers would be appreciated.

In the interests of more transparent development, [here's part of my test suite](https://github.com/Wheest/tvm-grouped-conv-test).

**Comment by FrozenGene** [comment]
> Hello there, updating this pull request to be up-to-date with the latest `main` branch.
> 
> In terms of things remaining to do:
> 
> * [x]  [Consider using compute_at and vectorize data load](https://github.com/apache/tvm/pull/6137#discussion_r463852066) - did not get an improvement.
> * [x]  [We should support asymmetic padding like other compute / schedule.](https://github.com/apache/tvm/pull/6137#pullrequestreview-459474121) - this is implemented in GSPC, however requires extending `get_workload` for Conv2D generally.  I began working on this in `505c127`, but have reverted it, and will have this as it's own pull request in the comings days.
> * [ ]  [Pack in alter_op_layout for kernel](https://github.com/apache/tvm/pull/6137#discussion_r463844394):  have been working on this, but have an issue.   My data is being passed to my `group_conv2d_NCHWc.x86` in the `conv2d_NCHWc` format (5D input data), rather than the GSPC format (6D input data).  Despite my changes to the x86 `_alter_conv2d_layout`.  [See this branch](https://github.com/Wheest/incubator-tvm/tree/wheest/gspc-dev-alter-op).  Some guidance or pointers would be appreciated @FrozenGene.
> 
> In the interests of more transparent development, [here's part of my test suite](https://github.com/Wheest/tvm-grouped-conv-test).

Sorry, @Wheest I just noticed your mention. Could you give me some spare time to review your code? I could allocate some buffer next week.

**Comment by Wheest** [comment]
Sure thing, thanks for reviewing!  You can see my ongoing PR for the asymmetric padding [here](https://github.com/apache/tvm/pull/7142)

**Comment by merrymercy** [comment]
We can also try the auto-scheduler. It should deliver very good performance for group convolution.

**Comment by Wheest** [comment]
Hi @merrymercy, I'm just reading about the auto-scheduler now, is the [RFC](https://github.com/apache/tvm/issues/2954) the best learning resource?  How does it compare to Ansor?

I'm willing to give it a try, though perhaps it would be better to get this PR merged first (since it's providing an immediate and clear performance improvement over the existing solution), and add improvements as another one?

**Comment by merrymercy** [comment]
The RFC you pointed is deprecated. "Ansor" and "Auto-scheduler" in the current code base are the same thing. They are just different names. You can see the RFC for Ansor auto-scheduler [here](https://discuss.tvm.apache.org/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005), but the RFC is also slightly old. The best thing to learn it is try these [tutorials](https://tvm.apache.org/docs/tutorials/index.html#autoscheduler-template-free-auto-scheduling).

**Comment by FrozenGene** [comment]
> here's part of my test suite

Not very clearly understand you problem. inside this pass, we should only handle kernel layout, not related with data layout. Could you explain it a bit more clearly?

File: python/tvm/relay/op/strategy/x86.py [hunk 5]
```diff
@@ -175,21 +175,10 @@ def conv2d_strategy_cpu(attrs, inputs, out_type, target):
...
-            assert kernel_layout == "HWIO"
-            if not is_auto_scheduler_enabled():
-                logger.warning("group_conv2d is not optimized for x86 with autotvm.")
             strategy.add_implementation(
-                wrap_compute_conv2d(topi.nn.group_conv2d_nhwc, has_groups=True),
-                wrap_topi_schedule(topi.generic.schedule_group_conv2d_nhwc),
-                name="group_conv2d_nhwc.generic",
+                wrap_compute_conv2d(topi.x86.group_conv2d_nchw, has_groups=True),
+                wrap_topi_schedule(topi.x86.schedule_group_conv2d_nchw),
+                name="group_conv2d_nchw.x86",
```
Comment (FrozenGene) [inline python/tvm/relay/op/strategy/x86.py:181]: We shouldn't delete auto scheduler NHWC support

Comment (FrozenGene) [inline python/tvm/relay/op/strategy/x86.py:181]: We shouldn't delete auto scheduler NHWC support

File: python/tvm/topi/x86/group_conv2d.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    ic_bn, oc_bn, reg_n, unroll_kw = (
+        cfg["tile_ic"].size[-1],
+        cfg["tile_oc"].size[-1],
+        cfg["tile_ow"].size[-1],
+        cfg["unroll_kw"].val,
+    )
+
+    # no stride and padding info here
+    padding = infer_pad(data, data_pad)
+    hpad, wpad = padding
```
Comment (FrozenGene) [inline python/tvm/topi/x86/group_conv2d.py:303]: Should this be `pad_t, pad_l, pad_b, pad_r`?

**Review by FrozenGene — CHANGES_REQUESTED** [review]

File: python/tvm/topi/x86/group_conv2d.py (hunk 6 above)
Comment (Wheest) [reply python/tvm/topi/x86/group_conv2d.py:303]: We only get `hpad` and `wpad` from [nn.utils.infer_pad](https://github.com/apache/tvm/blob/main/python/tvm/topi/nn/utils.py#L25), since this information can't be recovered from `data` and `data_pad` alone.  This has no effect on the schedule, since we only use it to get the `DOPAD` variable.

File: python/tvm/relay/op/strategy/x86.py [hunk 7]
```diff
@@ -175,21 +175,10 @@ def conv2d_strategy_cpu(attrs, inputs, out_type, target):
     else:  # group_conv2d
         if layout == "NCHW":
             assert kernel_layout == "OIHW"
-            if not is_auto_scheduler_enabled():
-                logger.warning("group_conv2d is not optimized for x86 with autotvm.")
-            strategy.add_implementation(
-                wrap_compute_conv2d(topi.nn.group_conv2d_nchw, has_groups=True),
-                wrap_topi_schedule(topi.generic.schedule_group_conv2d_nchw),
-                name="group_conv2d_nchw.generic",
-            )
```
[the remark this answers is not in the document]
Comment (Wheest) [reply python/tvm/relay/op/strategy/x86.py]: What kind of strategy logic would include both?  Having this code on an `else` branch of `if not is_auto_scheduler_enabled():`?  Or something else?

File: topi/python/topi/arm_cpu/group_conv2d.py (hunk 1 above)
Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py]: Regarding the issue I was having, I think I misunderstood, and was trying to alter the shape of both the data and the kernel ahead of time.  Am making the fix now.

Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py]: Hi @FrozenGene, my final blocker to completing this PR is adding the GSPC kernel layout to the C++ runtime.  I've got the Python side working, however the alter_op requires the layout to be available in the C++ runtime, and I'm unsure of how to do this.

See [this post on the forums](https://discuss.tvm.apache.org/t/adding-new-kernel-layout-to-backend/8824) where I explain my issue.  Would you be able to give any pointers please?

Comment (FrozenGene) [reply topi/python/topi/arm_cpu/group_conv2d.py]: @merrymercy @minminsun does this issue have the same reason as our auto scheduler layout rewrite issue? Do you have any idea about it?

Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py]: Hey @FrozenGene @merrymercy @minminsun, any thoughts on adding custom kernel layouts to the C++ runtime, so that `alter_op` can have kernels reshaped AoT?

Comment (FrozenGene) [reply topi/python/topi/arm_cpu/group_conv2d.py]: Yes. Sorry to delay this too long. We could open one issue to tracker it.

File: python/tvm/relay/op/strategy/x86.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (FrozenGene) [reply python/tvm/relay/op/strategy/x86.py]: `if not is_auto_scheduler_enabled()`: ... We prefer auto scheduler be the default stragegy.

File: python/tvm/topi/x86/group_conv2d.py (hunk 6 above)
Comment (FrozenGene) [reply python/tvm/topi/x86/group_conv2d.py:303]: I think we should fix the issue of `nn.utils.infer_pad` as we support asymmetric padding for now. Previous  `nn.utils.infer_pad` only return 2 values is the history reason. However, as it doesn't affect performance and correctness, it is up to you to decide whether to support it or not in this pr.

File: python/tvm/relay/op/strategy/x86.py [hunk 8]
```diff
@@ -205,22 +205,20 @@ def conv2d_strategy_cpu(attrs, inputs, out_type, target):
...
-            strategy.add_implementation(
-                wrap_compute_conv2d(topi.nn.group_conv2d_nhwc, has_groups=True),
-                wrap_topi_schedule(topi.generic.schedule_group_conv2d_nhwc),
-                name="group_conv2d_nhwc.generic",
-            )
+                strategy.add_implementation(
+                    wrap_compute_conv2d(topi.nn.group_conv2d_nhwc, has_groups=True),
+                    wrap_topi_schedule(topi.generic.schedule_group_conv2d_nhwc),
+                    name="group_conv2d_nhwc.generic",
+                )
```
Comment (FrozenGene) [inline python/tvm/relay/op/strategy/x86.py:221]: The fix has bug when we enable auto scheduler. When auto scheduler is enabled, we can not have any compute declaration for auto scheduler, please fix it.

**Review by FrozenGene — CHANGES_REQUESTED** [review]

File: python/tvm/topi/x86/group_conv2d.py (hunk 6 above)
Comment (Wheest) [reply python/tvm/topi/x86/group_conv2d.py:303]: Thanks, I think I will figure out `nn.utils.infer_pad` as a separate PR this weekend, since there are a couple of design issues around it that need to be answered.

File: python/tvm/relay/op/strategy/x86.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wheest) [reply python/tvm/relay/op/strategy/x86.py]: Current code should now be able to run with auto scheduler by running with `"relay.backend.use_auto_scheduler": True`, and otherwise using my schedule.

File: python/tvm/relay/op/strategy/x86.py (hunk 8 above)
Comment (Wheest) [reply python/tvm/relay/op/strategy/x86.py:221]: Bug has been fixed by removing the indentation.

File: topi/python/topi/arm_cpu/group_conv2d.py (hunk 1 above)
Comment (Wheest) [reply topi/python/topi/arm_cpu/group_conv2d.py]: Thanks.  I believe I have responded to the issues, and will see when the latest commit passes PR.  For the `nn.utils.infer_pad`, I will be making a separate PR, since this is a broader issue than GSPC.

**Comment by FrozenGene** [comment]
@Wheest Seems you have wrong rebase. please refer this documentation : https://tvm.apache.org/docs/contribute/git_howto.html

**Comment by Wheest** [comment]
> @Wheest Seems you have wrong rebase. please refer this documentation : https://tvm.apache.org/docs/contribute/git_howto.html

Believe I have fixed the rebase/commit problems, will try to avoid these in future for a cleaner PR.

**Review by FrozenGene — APPROVED** [review]

**Comment by FrozenGene** [comment]
Thanks @Wheest Merged now

**Comment by Wheest** [comment]
Thanks for helping me through my first PR for a major OSS project @FrozenGene.  Have learned a lot, and I hope to be a more productive member of the community in future.

**Comment by tqchen** [comment]
Thank you @Wheest ! and @FrozenGene for sheperding
